locationdelegation/src/main/java/com/google/androidbrowserhelper/locationdelegation/LocationProviderAndroid.java

```diff
@@ -0,0 +1,203 @@
+
+import java.util.List;
+
+import androidx.browser.trusted.TrustedWebActivityCallbackRemote;
+
+/**
+ * This is a LocationProvider using Android APIs [1]. It is a separate class for clarity
+ * so that it can manage all processing completely on the UI thread. The container class
+ * ensures that the start/stop calls into this class are done on the UI thread.
+ * <p>
+ * [1] https://developer.android.com/reference/android/location/package-summary.html
+ */
+public class LocationProviderAndroid implements LocationListener {
+    private static final String TAG = "TWA_LocationProvider";
+
+    private static final String EXTRA_NEW_LOCATION_AVAILABLE_CALLBACK = "onNewLocationAvailable";
+    private static final String EXTRA_NEW_LOCATION_ERROR_CALLBACK= "onNewLocationError";
+
+    private LocationManager mLocationManager;
+    private boolean mIsRunning;
+    private TrustedWebActivityCallbackRemote mCallback;
+    private static LocationProviderAndroid sProvider;
+
+    public static LocationProviderAndroid getInstance() {
+        if (sProvider == null) {
+            sProvider = new LocationProviderAndroid();
+        }
+        return sProvider;
+    }
+
+    public void start(
+            Context context, TrustedWebActivityCallbackRemote callback, boolean enableHighAccuracy) {
+        unregisterFromLocationUpdates();
+        mCallback = callback;
+        registerForLocationUpdates(context, enableHighAccuracy);
+    }
+
+    public void stop() {
+        unregisterFromLocationUpdates();
+    }
+
+    public boolean isRunning() {
+        return mIsRunning;
+    }
+
+    @Override
+    public void onLocationChanged(Location location) {
+        // Callbacks from the system location service are queued to this thread, so it's
+        // possible that we receive callbacks after unregistering. At this point, the
+        // native object will no longer exist.
+        if (mIsRunning) {
+            onNewLocationAvailable(location);
+        }
+    }
+
+    @Override
+    public void onStatusChanged(String provider, int status, Bundle extras) {
+    }
+
+    @Override
+    public void onProviderEnabled(String provider) {
+    }
+
+    @Override
+    public void onProviderDisabled(String provider) {
+    }
+
+    private void createLocationManagerIfNeeded(Context context) {
+        if (mLocationManager != null) return;
+        mLocationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
+        if (mLocationManager == null) {
+            Log.e(TAG, "Could not get location manager.");
+        }
+    }
+
+    /**
+     * Registers this object with the location service.
+     */
+    private void registerForLocationUpdates(Context context, boolean enableHighAccuracy) {
+        createLocationManagerIfNeeded(context);
+        if (usePassiveOneShotLocation()) return;
+
+        assert !mIsRunning;
+        mIsRunning = true;
+
+        // We're running on the main thread. The C++ side is responsible to
```

Comment:
The C++ side of what? Of Chrome? Of the Android location SDK?

Reply:
Removed now. This comments applies to the locationProvider class in chromium repo, not our version.